File: apps/api/app/schemas/municipal_registration.py
"""
Municipal Registration schemas for API requests and responses.
"""

from datetime import date, datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


class MunicipalRegistrationStatus(str, Enum):
    """Status of municipal registration"""
    ATIVA = "ativa"
    INATIVA = "inativa"
    SUSPENSA = "suspensa"
    PENDENTE = "pendente"
    CANCELADA = "cancelada"


class StateCode(str, Enum):
    """Brazilian state codes (UF)"""
    AC = "AC"
    AL = "AL"
    AP = "AP"
    AM = "AM"
    BA = "BA"
    CE = "CE"
    DF = "DF"
    ES = "ES"
    GO = "GO"
    MA = "MA"
    MT = "MT"
    MS = "MS"
    MG = "MG"
    PA = "PA"
    PB = "PB"
    PR = "PR"
    PE = "PE"
    PI = "PI"
    RJ = "RJ"
    RN = "RN"
    RS = "RS"
    RO = "RO"
    RR = "RR"
    SC = "SC"
    SP = "SP"
    SE = "SE"
    TO = "TO"


# Municipal Registration Schemas
class MunicipalRegistrationBase(BaseModel):
    """Base schema for municipal registration"""
    client_id: UUID
    city: str = Field(..., min_length=1, max_length=100, description="City name")
    state: StateCode = Field(..., description="State code (UF)")
    registration_number: str = Field(..., min_length=1, max_length=50, description="Municipal registration number (CCM/Inscrição Municipal)")
    issue_date: date = Field(..., description="Issue date")
    notes: Optional[str] = Field(None, max_length=1000, description="Additional notes")


class MunicipalRegistrationCreate(MunicipalRegistrationBase):
    """Schema for creating a new municipal registration"""
    pass


class MunicipalRegistrationUpdate(BaseModel):
    """Schema for updating a municipal registration"""
    city: Optional[str] = Field(None, min_length=1, max_length=100)
    state: Optional[StateCode] = None
    registration_number: Optional[str] = Field(None, min_length=1, max_length=50)
    issue_date: Optional[date] = None
    status: Optional[MunicipalRegistrationStatus] = None
    notes: Optional[str] = Field(None, max_length=1000)


class MunicipalRegistrationResponse(MunicipalRegistrationBase):
    """Schema for municipal registration response"""
    id: UUID
    status: MunicipalRegistrationStatus
    created_at: datetime
    updated_at: datetime

    # Related data
    client_name: Optional[str] = None

    class Config:
        from_attributes = True


class MunicipalRegistrationListResponse(BaseModel):
    """Schema for paginated municipal registration list"""
    items: list[MunicipalRegistrationResponse]
    total: int
    page: int
    size: int
    pages: int


# Statistics
class MunicipalRegistrationStatistics(BaseModel):
    """Municipal registration statistics"""
    total_registrations: int = 0
    active_registrations: int = 0
    inactive_registrations: int = 0
    by_state: dict[str, int] = Field(default_factory=dict)
    by_status: dict[str, int] = Field(default_factory=dict)

File: apps/web/src/types/municipal_registration.ts
/**
 * Municipal Registration types and interfaces
 */

// Enums
export enum MunicipalRegistrationStatus {
  ATIVA = "ativa",
  INATIVA = "inativa",
  SUSPENSA = "suspensa",
  PENDENTE = "pendente",
  CANCELADA = "cancelada",
}

export enum StateCode {
  AC = "AC",
  AL = "AL",
  AP = "AP",
  AM = "AM",
  BA = "BA",
  CE = "CE",
  DF = "DF",
  ES = "ES",
  GO = "GO",
  MA = "MA",
  MT = "MT",
  MS = "MS",
  MG = "MG",
  PA = "PA",
  PB = "PB",
  PR = "PR",
  PE = "PE",
  PI = "PI",
  RJ = "RJ",
  RN = "RN",
  RS = "RS",
  RO = "RO",
  RR = "RR",
  SC = "SC",
  SP = "SP",
  SE = "SE",
  TO = "TO",
}

// Labels
export const MUNICIPAL_REGISTRATION_STATUS_LABELS: Record<MunicipalRegistrationStatus, string> = {
  [MunicipalRegistrationStatus.ATIVA]: "Ativa",
  [MunicipalRegistrationStatus.INATIVA]: "Inativa",
  [MunicipalRegistrationStatus.SUSPENSA]: "Suspensa",
  [MunicipalRegistrationStatus.PENDENTE]: "Pendente",
  [MunicipalRegistrationStatus.CANCELADA]: "Cancelada",
};

export const STATE_LABELS: Record<StateCode, string> = {
  [StateCode.AC]: "Acre",
  [StateCode.AL]: "Alagoas",
  [StateCode.AP]: "Amapá",
  [StateCode.AM]: "Amazonas",
  [StateCode.BA]: "Bahia",
  [StateCode.CE]: "Ceará",
  [StateCode.DF]: "Distrito Federal",
  [StateCode.ES]: "Espírito Santo",
  [StateCode.GO]: "Goiás",
  [StateCode.MA]: "Maranhão",
  [StateCode.MT]: "Mato Grosso",
  [StateCode.MS]: "Mato Grosso do Sul",
  [StateCode.MG]: "Minas Gerais",
  [StateCode.PA]: "Pará",
  [StateCode.PB]: "Paraíba",
  [StateCode.PR]: "Paraná",
  [StateCode.PE]: "Pernambuco",
  [StateCode.PI]: "Piauí",
  [StateCode.RJ]: "Rio de Janeiro",
  [StateCode.RN]: "Rio Grande do Norte",
  [StateCode.RS]: "Rio Grande do Sul",
  [StateCode.RO]: "Rondônia",
  [StateCode.RR]: "Roraima",
  [StateCode.SC]: "Santa Catarina",
  [StateCode.SP]: "São Paulo",
  [StateCode.SE]: "Sergipe",
  [StateCode.TO]: "Tocantins",
};

// Base interfaces
export interface MunicipalRegistrationBase {
  client_id: string;
  city: string;
  state: StateCode;
  registration_number: string;
  issue_date: string; // ISO date string
  notes: string | null;
}

export interface MunicipalRegistrationCreate extends MunicipalRegistrationBase {}

export interface MunicipalRegistrationUpdate {
  city?: string;
  state?: StateCode;
  registration_number?: string;
  issue_date?: string;
  status?: MunicipalRegistrationStatus;
  notes?: string | null;
}

export interface MunicipalRegistration extends MunicipalRegistrationBase {
  id: string;
  status: MunicipalRegistrationStatus;
  created_at: string;
  updated_at: string;
  client_name?: string;
}

// List response
export interface MunicipalRegistrationListResponse {
  items: MunicipalRegistration[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

// Statistics
export interface MunicipalRegistrationStatistics {
  total_registrations: number;
  active_registrations: number;
  inactive_registrations: number;
  by_state: Record<string, number>;
  by_status: Record<string, number>;
}

// Filters
export interface MunicipalRegistrationFilters {
  client_id?: string;
  state?: StateCode;
  status?: MunicipalRegistrationStatus;
  search?: string; // Search in city, registration_number
  page?: number;
  size?: number;
}

// Helper functions
export function getMunicipalRegistrationStatusColor(
  status: MunicipalRegistrationStatus
): "success" | "warning" | "danger" | "default" {
  switch (status) {
    case MunicipalRegistrationStatus.ATIVA:
      return "success";
    case MunicipalRegistrationStatus.PENDENTE:
      return "warning";
    case MunicipalRegistrationStatus.INATIVA:
    case MunicipalRegistrationStatus.SUSPENSA:
    case MunicipalRegistrationStatus.CANCELADA:
      return "danger";
    default:
      return "default";
  }
}

export function getStateLabel(code: StateCode): string {
  return STATE_LABELS[code] || code;
}

export function getStateLabelWithCode(code: StateCode): string {
  return `${STATE_LABELS[code]} (${code})`;
}
